config/parse.go

```diff
@@ -134,80 +96,125 @@ func (ctx *parseContext) isSnippet(name string) (bool, string) {
-			if ctx.parens < 0 {
-				return res, ctx.Err("Unexpected }")
+			ctx.nesting--
+			// name arg0 arg1 { #<1
+			// }   }
+			// ^2  ^3
+			//
+			// After #1 ctx.nesting is incremented by ctx.nesting++ before this loop.
+			// Then we advance cursor and hit }, we exit loop, ctx.nesting now becomes 0.
+			// But then parent block reader does the same when it hits #3 -
+			// ctx.nesting becomes -1 and it fails.
+			if ctx.nesting < 0 {
+				return res, ctx.Err("unexpected }")
 			}
 			break
 		}
-		node, err := ctx.readCfgNode()
+		node, err := ctx.readNode()
 		if err != nil {
 			return res, err
 		}
 
 		shouldStop := false
+
+		// name arg0 arg1 {
+		//   c1 c2 }
+		//         ^
+		// Edgy case, here we check if last argument of last node is a }
+		// If it is - we stop as we hit end of our block.
 		if len(node.Args) != 0 && node.Args[len(node.Args)-1] == "}" {
-			ctx.parens--
-			if ctx.parens < 0 {
-				return res, ctx.Err("Unexpected }")
+			ctx.nesting--
+			if ctx.nesting < 0 {
+				return res, ctx.Err("unexpected }")
 			}
 			node.Args = node.Args[:len(node.Args)-1]
 			shouldStop = true
 		}
 
 		if node.Snippet {
-			if ctx.parens != 0 {
-				return res, ctx.Err("Snippet declarations are only allowed at top-level")
+			if ctx.nesting != 0 {
+				return res, ctx.Err("snippet declarations are only allowed at top-level")
 			}
 			if len(node.Args) != 0 {
-				return res, ctx.Err("Snippet declarations can't have arguments")
+				return res, ctx.Err("snippet declarations can't have arguments")
 			}
 
 			ctx.snippets[node.Name] = node.Children
 			continue
 		}
 
+		res = append(res, node)
 		if shouldStop {
 			break
 		}
-		res = append(res, node)
 	}
 	return res, nil
 }
 
-func readCfgTree(r io.Reader, location string, depth int) (nodes []Node, snips map[string][]Node, err error) {
+func readTree(r io.Reader, location string, expansionDepth int) (nodes []Node, snips map[string][]Node, err error) {
 	ctx := parseContext{
 		Dispenser:    caddyfile.NewDispenser(location, r),
 		snippets:     make(map[string][]Node),
+		nesting:      -1,
 		fileLocation: location,
 	}
+
 	root := Node{}
-	root.Children, err = ctx.readCfgNodes()
+	root.File = location
+	root.Line = 1
+	// Before parsing starts lexer's cursor points to the non-existent token before
+	// first one. From readNodes viewpoint this is opening brace so we
+	// don't break any requirements here.
+	//
+	// For the same reason we use -1 as a starting nesting. So readNodes
+	// will see this as it is reading block at nesting level 0.
+	root.Children, err = ctx.readNodes()
 	if err != nil {
 		return root.Children, ctx.snippets, err
 	}
-	if ctx.parens < 0 {
-		return root.Children, ctx.snippets, ctx.Err("Unexpected }")
-	}
-	if ctx.parens > 0 {
-		return root.Children, ctx.snippets, ctx.Err("Unexpected {")
+
+	// There is no need to check ctx.nesting < 0 because it is checked by readNodes.
+	if ctx.nesting > 0 {
+		return root.Children, ctx.snippets, ctx.Err("unexpected {")
```

Comment:
Isn't this "expected }" instead?

Reply:
`ctx.nesting` is incremented every time we parse `{` and decremented every time we parse `}`.
So if after parsing `ctx.nesting` is negative - we have more }'s than {'s, if `ctx.nesting` is positive - we have more {'s than }'s. 
Though this error should be corrected to be "unexpected EOF" because it means we hit EOF when we were looking for }, /u didn't notice that when moving code around.

Reply:
Changed error description (64ef51c).